### PR TITLE
Adding support for JWT aud claims

### LINKED ISSUE #19: Error decoding a token with audience claim
I'm getting a "Invalid audience" error from the pyramid_jwt lib. There is an audience section in the token:
```
{
"nbf": 1520083317,
"exp": 1520086917,
"iss": "https://demo.identityserver.io",
"aud": [
"https://demo.identityserver.io/resources",
"api"
],
"client_id": "client",
"scope": [
"api"
]
}
```
Can I somehow pass an audience param down to the decode method?

### PR DESCRIPTION
**Pull request by crooksey**
Fixes #19 

Code written, tests added.

Would appreciate someone else to test before merging, works for me tested multiple setups.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -157,6 +167,9 @@ You can either set this in your .ini-file, or pass/override them directly to the
 | expiration   | jwt.expiration  |               | Number of seconds (or a datetime.timedelta |
 |              |                 |               | instance) before a token expires.          |
 +--------------+-----------------+---------------+--------------------------------------------+
+| audience     | jwt.audience    |               | Proposed audience for the token            |
+|              |                 |               | instance) before a token expires.          |
```
Comment (wichert) [inline README.rst]: There is a stray `)` there

File: setup.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 tests_require = [
     'pytest',
-    'WebTest',
+    'WebTest'
```
Comment (wichert) [inline setup.py]: Please leave that comma in place.

File: src/pyramid_jwt/__init__.py [hunk 3]
```diff
@@ -9,7 +9,7 @@ def includeme(config):
 
 
 def create_jwt_authentication_policy(config, private_key=None, public_key=None,
-        algorithm=None, expiration=None, leeway=None,
+        algorithm=None, expiration=None, leeway=None, audience=None,
```
Comment (wichert) [inline src/pyramid_jwt/__init__.py]: I am a bit worried about adding a parameter in the middle here: theoretically you break backwards compatibility of someone does not use named parameters.

File: src/pyramid_jwt/__init__.py [hunk 4]
```diff
@@ -32,22 +32,23 @@ def create_jwt_authentication_policy(config, private_key=None, public_key=None,
...
             http_header=http_header,
             auth_type=auth_type,
             callback=callback,
             json_encoder=json_encoder)
 
 
 def set_jwt_authentication_policy(config, private_key=None, public_key=None,
-        algorithm=None, expiration=None, leeway=None,
+        algorithm=None, expiration=None, leeway=None, audience=None,
         http_header=None, auth_type=None, callback=None, json_encoder=None):
```
Comment (wichert) [inline src/pyramid_jwt/__init__.py]: Same as earlier:I am a bit worried about adding a parameter in the middle here: theoretically you break backwards compatibility of someone does not use named parameters.

File: src/pyramid_jwt/policy.py [hunk 5]
```diff
@@ -61,12 +68,24 @@ def get_claims(self, request):
...
+                            request.remote_addr, e)
+                return {}
+        else:
+            try:
+                claims = jwt.decode(token, self.public_key, algorithms=[self.algorithm],
+                                    leeway=self.leeway)
+                return claims
+            except jwt.InvalidTokenError as e:
+                log.warning('Invalid JWT token from %s: %s', request.remote_addr, e)
+                return {}
```
Comment (wichert) [inline src/pyramid_jwt/policy.py]: That code duplication looks bad. Does PyJWT do something unexpected if you pass in `audience=None` to require that? If so I suggest a different approach:

```python
extra = {}
if audience:
    extra['audience'] = audience
try:
   claims = jwt.decode(token, …, **extra)
except jwt.InvalidTokenError as e:
    …
```

Comment (crooksey) [reply src/pyramid_jwt/policy.py]: It was more to give a specific warning log that the decode failed due to an unmatched audience, without the code duplication the error is not as helpful to the user. How does this look:
```python
audience = self.audience
        def decode_jwt(self, token):
            claims = jwt.decode(token, self.public_key, algorithms=[self.algorithm],
                                    audience=self.audience, leeway=self.leeway)
            return claims

        if audience:
            try:
                decode_jwt(self, token=token)
            except jwt.InvalidTokenError as e:
                log.warning('Invalid matched audience for JWT token from %s: %s', 
                            request.remote_addr, e)
                return {}
        else:
            try:
                decode_jwt(self, token=token)
            except jwt.InvalidTokenError as e:
                log.warning('Invalid JWT token from %s: %s', request.remote_addr, e)
                return {}
```

Comment (wichert) [reply src/pyramid_jwt/policy.py]: Why not do the `if` in the `except` just to wrap the `log.warning` call?

```python
try:
   claims = jwt.decode(token, …, audience=audience)
except jwt.InvalidTokenError as e:
    if audience:
        log.warning('Invalid matched audience')
    else:
        log.warning('Invalid JWT token')
```

Comment (crooksey) [reply src/pyramid_jwt/policy.py]: That's a good idea, I will update and test this on Monday.

**Comment by jstevensfit** [comment]
@wichert when will this be released on pypi?

**Comment by wichert** [comment]
@jstevensfit I've just made a release!

**Comment by jstevensfit** [comment]
@wichert Thanks for the quick response, it's working great.

**Pull request merged**